pkg/controller/s3/s3bucketpolicy/s3bucketpolicy.go

```diff
@@ -0,0 +1,222 @@
+*/
+
+package s3bucketpolicy
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	awss3 "github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/google/go-cmp/cmp"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	runtimev1alpha1 "github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	bucketv1alpha3 "github.com/crossplane/provider-aws/apis/storage/v1alpha3"
+	"github.com/crossplane/provider-aws/pkg/clients/iam"
+	"github.com/crossplane/provider-aws/pkg/clients/s3"
+	"github.com/crossplane/provider-aws/pkg/controller/utils"
+)
+
+const (
+	errUnexpectedObject = "The managed resource is not an IAMRolePolicyAttachment resource"
+	errClient           = "cannot create a new RolePolicyAttachmentClient"
+	errAttach           = "failed to attach the policy to role"
+	errDelete           = "failed to delete the policy for bucket"
+	errGet              = "failed to get S3BucketPolicy for bucket with name"
+	errUpdate           = "failed to update the policy for bucket"
+)
+
+//var (
+//ctx           = context.Background()
+//result        = reconcile.Result{}
+//resultRequeue = reconcile.Result{Requeue: true}
+//)
+
+// SetupS3BucketPolicy adds a controller that reconciles
+// S3BucketPolicies.
+func SetupS3BucketPolicy(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(bucketv1alpha3.S3BucketPolicyGroupKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&bucketv1alpha3.S3BucketPolicy{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(bucketv1alpha3.S3BucketPolicyGroupVersionKind),
+			managed.WithExternalConnecter(&connector{client: mgr.GetClient(), newClientFn: s3.NewBucketPolicyClient, awsConfigFn: utils.RetrieveAwsConfigFromProvider}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithInitializers(managed.NewNameAsExternalName(mgr.GetClient())),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connector struct {
+	client      client.Client
+	newClientFn func(*aws.Config) (s3.BucketPolicyClient, iam.Client, error)
+	awsConfigFn func(context.Context, client.Reader, *corev1.ObjectReference) (*aws.Config, error)
+}
+
+func (conn *connector) Connect(ctx context.Context, mgd resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mgd.(*bucketv1alpha3.S3BucketPolicy)
+	if !ok {
+		return nil, errors.New(errUnexpectedObject)
+	}
+
+	awsconfig, err := conn.awsConfigFn(ctx, conn.client, cr.Spec.ProviderReference)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Would be good to try to use newer pattern here, but not critical (example: https://github.com/crossplane/provider-aws/blob/778a71cc241103b48ea4ff3d12bf0aea7a92bfb7/pkg/controller/eks/cluster.go#L89)